pkg/apiserver/internal/v1/offering.go

```diff
@@ -21,67 +21,133 @@ import (
-	return &offeringServer{
-		client: c,
+func NewOfferingServiceServer(c client.Client, dynamicClient dynamic.Interface, restMapper meta.RESTMapper, scheme *runtime.Scheme) (v1.OfferingServiceServer, error) {
+	offeringServer := &offeringServer{
+		client:        c,
+		dynamicClient: dynamicClient,
+		scheme:        scheme,
+	}
+	objGVK, err := apiutil.GVKForObject(&catalogv1alpha1.Offering{}, offeringServer.scheme)
+	if err != nil {
+		return nil, err
+	}
+	restMapping, err := restMapper.RESTMapping(objGVK.GroupKind(), objGVK.Version)
+	if err != nil {
+		return nil, err
 	}
+	offeringServer.gvr = restMapping.Resource
+	return offeringServer, nil
 }
 
 func (o offeringServer) List(ctx context.Context, req *v1.OfferingListRequest) (res *v1.OfferingList, err error) {
 	var listOptions []client.ListOption
 	listOptions, err = o.validateListRequest(req)
 	if err != nil {
-		return nil, status.Errorf(codes.InvalidArgument, err.Error())
+		return nil, status.Error(codes.InvalidArgument, err.Error())
 	}
 	offeringList := &catalogv1alpha1.OfferingList{}
 	if err := o.client.List(ctx, offeringList, listOptions...); err != nil {
-		return nil, status.Errorf(codes.Internal, fmt.Sprintf("listing offerings: %s", err.Error()))
+		return nil, status.Errorf(codes.Internal, "listing offerings: %s", err.Error())
 	}
 
 	res, err = o.convertOfferingList(offeringList)
 	if err != nil {
-		return nil, status.Errorf(codes.Internal, fmt.Sprintf("converting OfferingList: %s", err.Error()))
+		return nil, status.Errorf(codes.Internal, "converting OfferingList: %s", err.Error())
 	}
 	return
 }
 
 func (o offeringServer) Get(ctx context.Context, req *v1.OfferingGetRequest) (res *v1.Offering, err error) {
 	if err = o.validateGetRequest(req); err != nil {
-		return nil, status.Errorf(codes.InvalidArgument, err.Error())
+		return nil, status.Error(codes.InvalidArgument, err.Error())
 	}
 	offering := &catalogv1alpha1.Offering{}
 	if err = o.client.Get(ctx, types.NamespacedName{
 		Name:      req.Name,
 		Namespace: req.Account,
 	}, offering); err != nil {
-		return nil, status.Errorf(codes.Internal, fmt.Sprintf("getting offering: %s", err.Error()))
+		return nil, status.Errorf(codes.Internal, "getting Offering: %s", err.Error())
 	}
 	res, err = o.convertOffering(offering)
 	if err != nil {
-		return nil, status.Errorf(codes.Internal, fmt.Sprintf("converting Offering: %s", err.Error()))
+		return nil, status.Errorf(codes.Internal, "converting Offering: %s", err.Error())
 	}
 	return
 }
 
+func (o offeringServer) Watch(req *v1.OfferingWatchRequest, stream v1.OfferingService_WatchServer) error {
+	listOptions, err := o.validateWatchRequest(req)
+	if err != nil {
+		return status.Error(codes.InvalidArgument, err.Error())
+	}
+	watcher, err := o.dynamicClient.Resource(o.gvr).Namespace(req.Account).Watch(listOptions)
+	if err != nil {
+		return status.Errorf(codes.Internal, "watching offerings: %s", err.Error())
+	}
+	for {
```

Comment:
`defer watcher.Stop()` to prevent resource leakage.

Reply:
done